lib/philomena/autocomplete/generator.ex

```diff
@@ -5,32 +5,56 @@ defmodule Philomena.Autocomplete.Generator do
   See assets/js/utils/local-autocompleter.ts for how this should be used.
   The file follows the following binary format:
 
+      // Note that all pointer types (`struct T*`) are all relative to the binary payload start.
+
+      // Variable-length struct.
       struct tag {
-        uint8_t key_length;
-        uint8_t key[];
-        uint8_t association_length;
+        // Full name of the tag in UTF8.
+        uint8_t name_length;
+        uint8_t name[];
+
+        // List of IDs of other tags that appear on the same images in >50% of cases
+        uint8_t associations_length;
         uint32_t associations[];
       };
 
+      // Fixed-length struct.
       struct tag_reference {
-        uint32_t tag_location;
-        union {
-          int32_t raw;
-          uint32_t num_uses;    ///< when positive
-          uint32_t alias_index; ///< when negative, -alias_index - 1
-        };
+        // 32 bit pointer to a tag in the `tags` array
+        struct tag* tag;
+
+        // If >=0 then this tag is canonical and the `meta` is the number of images with this tag
+        //
+        // If <0 then the following calculation: `-meta - 1` and will give the index
```

Comment:
may be helpful to include the design reason for this
```suggestion
        // If <0 then `-meta - 1` (to allow 0 as a possibility for the target) will give the index
```